sstables_loader.cc

```diff
@@ -115,75 +115,217 @@ class send_meta_data {
+    future<> stream_sstable_mutations(const dht::partition_range&, std::vector<sstables::shared_sstable>, bool primary_replica_only);
+};
+
+class tablet_sstable_streamer : public sstable_streamer {
+    const locator::tablet_map& _tablet_map;
+public:
+    tablet_sstable_streamer(netw::messaging_service& ms, replica::database& db, ::table_id table_id, std::vector<sstables::shared_sstable> sstables)
+        : sstable_streamer(ms, db, table_id, std::move(sstables))
+        , _tablet_map(_erm->get_token_metadata().tablets().get_tablet_map(table_id)) {
+    }
+
+    virtual future<> stream(bool primary_replica_only) override;
+    virtual inet_address_vector_replica_set get_endpoints(const dht::token& token, bool primary_replica_only) const override;
+private:
+    future<> stream_fully_contained_sstables(const dht::partition_range& pr, std::vector<sstables::shared_sstable> sstables,
+                                             bool primary_replica_only) {
+        // FIXME: fully contained sstables can be optimized.
+        return stream_sstables(pr, std::move(sstables), primary_replica_only);
+    }
+};
+
+inet_address_vector_replica_set sstable_streamer::get_endpoints(const dht::token& token, bool primary_replica_only) const {
+    auto current_targets = _erm->get_natural_endpoints(token);
+    if (primary_replica_only && current_targets.size() > 1) {
+        current_targets.resize(1);
+    }
+    return current_targets;
+}
+
+inet_address_vector_replica_set tablet_sstable_streamer::get_endpoints(const dht::token& token, bool primary_replica_only) const {
+    auto current_targets = sstable_streamer::get_endpoints(token, primary_replica_only);
+
+    auto pending_replica_endpoint = std::invoke([&] () -> std::optional<gms::inet_address> {
+        auto tid = _tablet_map.get_tablet_id(token);
+        auto* tinfo = _tablet_map.get_tablet_transition_info(tid);
+        // No tablet in transit.
+        if (!tinfo) {
+            return std::nullopt;
+        }
+
+        auto leaving_replica = locator::get_leaving_replica(_tablet_map.get_tablet_info(tid), *tinfo);
+        auto& tm = _erm->get_token_metadata();
+        auto is_leaving = [&] (gms::inet_address target) {
+            return tm.get_host_id(target) == leaving_replica.host;
+        };
+        // If primary_replica_only is set, then it might happen the primary replica is not
+        // the one leaving through migration. Or migration is at such an advanced stage,
+        // that the next replica set is picked and none are leaving.
+        if (std::ranges::none_of(current_targets, is_leaving)) {
+            return std::nullopt;
+        }
+        // If intra-node migration is happening, we don't want to stream to same node twice.
+        if (leaving_replica.host == tinfo->pending_replica.host) {
+            return std::nullopt;
+        }
+        return tm.get_endpoint_for_host_id(tinfo->pending_replica.host);
+    });
+    if (pending_replica_endpoint) {
```

Comment:
Consider:

Old replicas: n1, n2, n3
New replicas: n1, n2, n4

With primary_replica_only = true, with this patch, sstable_streamer::get_endpoints would return n1. tablet_sstable_streamer::get_endpoints would return n1 and n4.  This is still not respecting primary_replica_only. 
Since the new replica node n4 is not the primary replica node, even if it is a pending node, we should not stream to it. tablet_sstable_streamer::get_endpoints should return n1.

Reply:
Actually it wouldn't. See tablet version uses base one for getting natural endpoints. So n1,n2 and n3 is trimmed to n1. Then we realize n4 is not replacing n1 but n3 instead. So we just return n1

Reply:
I am not at computer now but you can verify what I said by looking at none_of... It checks if none of the current ranges is leaving (in the example with primary replica only set, it would be n1) and if so we won't include pending replica == n4.

Reply:
You are right. The none_of check is making sure we do not return n4 in this case.